### PR TITLE
Move deploy and update into the plugin for ease of unittesting

### PR DESCRIPTION
**Pull request by asalkeld**
This enables us to test the deploy and update flows with simple mocking instead of within cluster.

### CODE REVIEW

File: pkg/plugin/plugin_test.go [hunk 1]
```diff
@@ -20,211 +20,52 @@ func TestCreateOrUpdate(t *testing.T) {
...
-			name:     "deploy: deploy error",
-			isUpdate: false,
-			wantErr:  true,
-			errStep:  api.PluginStepDeploy,
-		},
-		{
-			name:     "deploy: initialize error",
-			isUpdate: false,
-			wantErr:  true,
-			errStep:  api.PluginStepInitialize,
```
Comment (asalkeld) [inline pkg/plugin/plugin_test.go]: These tests were not very useful, so removing. the mocking below confirms the steps that we want to take for each flow.

File: pkg/cluster/healthcheck.go [hunk 2]
```diff
@@ -68,3 +68,8 @@ func (u *simpleUpgrader) HealthCheck(ctx context.Context, cs *api.OpenShiftManag
 	u.log.Info("checking admin console health")
 	return u.doHealthCheck(ctx, getHealthCheckHTTPClient(cs), "https://console."+cs.Properties.RouterProfiles[0].PublicSubdomain, 10*time.Second)
 }
+
+func (u *simpleUpgrader) WaitForHTTPStatusOk(ctx context.Context, cs *api.OpenShiftManagedCluster) error {
```
Comment (mjudeikis) [inline pkg/cluster/healthcheck.go]: This name is confusing now. It is checking master healthz endpoint, but name is not define what we are checking :man_shrugging:

**Comment by mjudeikis** [comment]
lgtm, will wait for next reviewer too

File: pkg/cluster/healthcheck.go (hunk 2 above)
Comment (kwoodson) [reply pkg/cluster/healthcheck.go]: Probably makes sense to update this or create a follow up (issue/PR) that does.

File: pkg/plugin/plugin_test.go [hunk 3]
```diff
@@ -20,211 +21,52 @@ func TestCreateOrUpdate(t *testing.T) {
...
+			AgentPoolProfiles: []api.AgentPoolProfile{
+				{Role: api.AgentPoolProfileRoleMaster, Name: "master"},
+				{Role: api.AgentPoolProfileRoleCompute, Name: "compute"},
+				{Role: api.AgentPoolProfileRoleInfra, Name: "infra"},
+			},
 		},
 	}
+
 	for _, tt := range tests {
 		mockGen := mock_arm.NewMockGenerator(mockCtrl)
```
Comment (jim-minter) [inline pkg/plugin/plugin_test.go]: it's a complete nit but I wonder if it's worth trying to standardise in the direction of calling the mocks by the fieldname they go into, ie s/mockGen/armGenerator/, s/mockUp/clusterUpgrader/. See update_master_test.go as an example.

File: pkg/plugin/plugin_test.go [hunk 4]
```diff
@@ -20,211 +21,52 @@ func TestCreateOrUpdate(t *testing.T) {
...
-				} else {
-					mockUp.EXPECT().Deploy(nil, nil, nil, nil, gomock.Any()).Return(nil)
-				}
-				mockUp.EXPECT().WaitForInfraServices(nil, nil).Return(err)
-			}
+		mockGen.EXPECT().Generate(nil, cs, "", tt.isUpdate, gomock.Any()).Return(nil, nil)
+		mockUp.EXPECT().CreateClients(nil, cs).Return(nil)
+		mockUp.EXPECT().Initialize(nil, cs).Return(nil)
+		if tt.isUpdate {
+			mockUp.EXPECT().UpdateMasterAgentPool(nil, cs, &cs.Properties.AgentPoolProfiles[0]).Return(nil)
```
Comment (jim-minter) [inline pkg/plugin/plugin_test.go]: I'm interested to know whether it's plausible to strengthen this as in:
```golang
c := foo.EXPECT().Bar().Return(baz)
c = bar.EXPECT().Baz().Return(bash).After(c)
c = baz.EXPECT().Bash().Return(qux).After(c)
```
(again as in update_master_test.go)
or whether we think it'll make us go mad.  Thoughts?

File: pkg/cluster/healthcheck.go [hunk 5]
```diff
@@ -68,3 +68,8 @@ func (u *simpleUpgrader) HealthCheck(ctx context.Context, cs *api.OpenShiftManag
 	u.log.Info("checking admin console health")
 	return u.doHealthCheck(ctx, getHealthCheckHTTPClient(cs), "https://console."+cs.Properties.RouterProfiles[0].PublicSubdomain, 10*time.Second)
 }
+
+func (u *simpleUpgrader) WaitForHealthzStatusOk(ctx context.Context, cs *api.OpenShiftManagedCluster) error {
+	_, err := wait.ForHTTPStatusOk(ctx, u.log, u.rt, "https://"+cs.Properties.FQDN+"/healthz")
```
Comment (jim-minter) [inline pkg/cluster/healthcheck.go]: can you think about whether it's worth removing `u.rt` at this point if the layering is better?  not sure either way - just have a think about it

File: pkg/plugin/plugin_test.go (hunk 3 above)
Comment (asalkeld) [reply pkg/plugin/plugin_test.go]: done

File: pkg/plugin/plugin_test.go (hunk 4 above)
Comment (asalkeld) [reply pkg/plugin/plugin_test.go]: done, easy enough

File: pkg/cluster/healthcheck.go (hunk 5 above)
Comment (asalkeld) [reply pkg/cluster/healthcheck.go]: we *could*, but after trying it, it makes the plugin_test.go messy

**Pull request merged**